types/node.lua

```diff
@@ -204,17 +177,34 @@ function nodetypes.hbox:outputYourself (typesetter, line)
    end
 end
 
-nodetypes.zerohbox = pl.class(nodetypes.hbox)
-nodetypes.zerohbox.type = "zerohbox"
-nodetypes.zerohbox.value = { glyph = 0 }
-
-nodetypes.nnode = pl.class(nodetypes.hbox)
-nodetypes.nnode.type = "nnode"
-nodetypes.nnode.language = ""
-nodetypes.nnode.pal = nil
-nodetypes.nnode.nodes = {}
-
-function nodetypes.nnode:_init (spec)
+--- A zerohbox is a special-kind of hbox with zero width, height and depth.
+--
+-- Derived from `hbox`.
+--
+-- Properties is_zerohbox (and convenience is_zero) and is_box are true.
+-- Note that is_hbox is NOT true: zerohbox are used in a specific context
+--
+-- @type zerohbox
+local zerohbox = pl.class(hbox)
+zerohbox.type = "zerohbox"
+zerohbox.value = { glyph = 0 }
+
+--- A nnode is a node representing text content.
+--
+-- Derived from `hbox`.
+--
+-- Properties is_nnode and is_box are true.
+--
+-- @type nnode
+local nnode = pl.class(hbox)
+nnode.type = "nnode"
+nnode.language = ""
+nnode.pal = nil
```

Comment:
Let me die in utter pain :rofl: What is this? Ahhh might have found it. The Pango shaper again! (See another comment relating to it.) OMG.

Reply:
So the next review we should do is the shapers, to see if the API can be made more generic, no?

If we *can't* avoid enriching/annotating (n)nodes with shaper-specific stuff, at least we ought to put them in a subobject. `nnode.pal` --> `nnode._shaper_specific.pango.pal` or anything similar so one _knows_ that the data is somewhat internal and tied to a shaper... for reasons...

Reply:
Are there even builds of SILE with the pango shaper? 'Cause keeping a software maintainable is also the take, sometimes, the decision to cut some branches in the tree, if they never blossomed.
I am not even sure it is useful as an "example" of anything, since it probably breaks everywhere it can.